pkg/storegateway/indexheader/index/symbols_test.go

```diff
@@ -53,37 +178,81 @@ func TestSymbols(t *testing.T) {
 	// We store only 4 offsets to symbols.
 	require.Len(t, s.offsets, 4)
 
-	for i := 99; i >= 0; i-- {
-		s, err := s.Lookup(uint32(i))
+	t.Run("Lookup", func(t *testing.T) {
+		for i := 99; i >= 0; i-- {
+			s, err := s.Lookup(uint32(i))
+			require.NoError(t, err)
+			require.Equal(t, string(rune(i)), s)
+		}
+		_, err = s.Lookup(100)
+		require.Error(t, err)
+	})
+
+	t.Run("ReverseLookup", func(t *testing.T) {
+		for i := 99; i >= 0; i-- {
+			r, err := s.ReverseLookup(string(rune(i)))
+			require.NoError(t, err)
+			require.Equal(t, uint32(i), r)
+		}
+		_, err = s.ReverseLookup(string(rune(100)))
+		require.Error(t, err)
+	})
+
+	t.Run("ForEachSymbol", func(t *testing.T) {
+		// Use ForEachSymbol to build an offset -> symbol mapping and ensure
+		// that it matches the expected offsets and symbols.
+		var symbols []string
+		expected := make(map[uint32]string)
+		for i := 99; i >= 0; i-- {
+			symbols = append(symbols, string(rune(i)))
+			expected[uint32(i)] = string(rune(i))
+		}
+
+		actual := make(map[uint32]string)
+		err = s.ForEachSymbol(symbols, func(sym string, offset uint32) error {
+			actual[offset] = sym
+			return nil
+		})
+
 		require.NoError(t, err)
-		require.Equal(t, string(rune(i)), s)
-	}
-	_, err = s.Lookup(100)
-	require.Error(t, err)
+		require.Equal(t, expected, actual)
+	})
+
+	t.Run("Reader iterate all", func(t *testing.T) {
+		r := s.Reader()
 
-	for i := 99; i >= 0; i-- {
-		r, err := s.ReverseLookup(string(rune(i)))
+		for i := uint32(0); i >= 99; i++ {
```

Comment:
I'm still very confused by this test. Shouldn't be:

```suggestion
		for i := uint32(0); i <= 99; i++ {
```

Reply:
ah, that's right, I had https://github.com/grafana/mimir/pull/5073/commits/750985421acc42a95cdb1de2d134624317335863. fixed it in, but then reset the changes that weren't included in 5103